### ISSUE TITLE
[New Game Request] Tang Tang (MAME 2010 Full Non-Merged Reference Set)

### ISSUE THREAD

**Issue by SapphireTactician** [libretro/mame2003-plus-libretro#148]
### Description

@arcadez @markwkidd There's a game from the MAME 2010 Reference Set that I'd really like to see playable on the MAME 2003 Plus core, called "Tang Tang" (*rom name: `tangtang`*). This game is not part of the MAME 2003 set as the game was first dumped in November 2005 and it's first MAME release was on December 1st, 2005 for Mame 0.102u2.

A MAME 2010 core / FBA core would be too slow for the devices I'd like to use this game on. Please make my dream of playing this game on MAME 2003 Plus come true.

-----

### Additional Info:

@markwkidd Advised me that I can help the process of support for this game being added by providing the following information from: http://adb.arcadeitalia.net/dettaglio_mame.php?game_name=tangtang&lang=en

**First Release:** Mame 0.102u2 released on Dec-01 2005

**Info Readout:**

* 0.182: David Haywood added a preconfigured NVRAM to Tang Tang so it won't boot with invalid settings.

* 0.102u4: Pierpaolo Prazzoli fixed sprite priorities in Tang Tang. Added missing gfx2 roms. Changed description to 'Tang Tang (ver. 0526, 26/05/2000)'.

* 13th December 2005: David Haywood - f205v of the MAME Italia Forum finished off Tang Tang and Spectrum 2000 by adding loading for the correct mask roms.

* 0.102u2: David Haywood added 'Tang Tang' (ESD 2000).

* 27th November 2005: David Haywood - Added Tang Tang to the esd16 driver, the GFX roms containing the backgrounds aren't dumped.

* 25th November 2005: f205v dumped Tang Tang (ver. 0526, 26/05/2000).

**Early Driver History:**

* 0.137: David Haywood implemented default EEPROMs to ESD16 driver.
* 0.135u2: Added includes\esd16.h. Fabio Priuli added driver data and save state support to ESD16 driver.
* 0.131u1: MooglyGuy merged memory maps in the ESD16 driver.
* 0.111u5: Brian Troha added documentation to the ESD16 driver.
* 0.102: Pierpaolo Prazzoli added Tilemap0 color banking to ESD16 driver.
* 0.37b16: Added esd16.c driver and vidhrdw\esd16.c.

**Comment by arcadez** [comment]
Alright im on the case apart from Tang Tang another game called Head Panic would be
playable also with the driver update

Regards

**Comment by arcadez** [comment]
well gimmie a chance bigman all the prelim work is done i now have the very tedious and
laborious task of rolling back all the memory maps to MAME78 standards then test all the
games.

I should have a new commit ready later today or tommorow.

**Comment by arcadez** [comment]
Your request has been fulfilled ;)

**Comment by SapphireTactician** [comment]
@arcadez I couldn't get the build needed in the Core Updater, so I attempted to compile it myself and got the following error:

```
src/vidhrdw/esd16_vidhrdw.c:44:1: error: C++ style comments are not allowed in ISO C90
 //extern data16_t *head_unknown1;
 ^
src/vidhrdw/esd16_vidhrdw.c:44:1: error: (this will be reported only once per input file)
Makefile:493: recipe for target 'src/vidhrdw/esd16_vidhrdw.o' failed
make: *** [src/vidhrdw/esd16_vidhrdw.o] Error 1
```

Later builds stop at this error instead:

```
src/mame2003/mame2003.c:1042:3: note: in expansion of macro ‘EMIT_RETRO_PAD_DIRECTIONS’
   EMIT_RETRO_PAD_DIRECTIONS(1),(options.retropad_layout = "modern" : EMIT_RETRO_PAD_MODERN(1) ? (options.retropad_layout = "snes" : EMIT_RETRO_PAD_SNES(1) ? EMIT_RETRO_PAD(1))),EMIT_RETRO_PAD_MOUSE(1),
   ^~~~~~~~~~~~~~~~~~~~~~~~~
src/mame2003/mame2003.c:1042:31: error: expected expression before ‘,’ token
   EMIT_RETRO_PAD_DIRECTIONS(1),(options.retropad_layout = "modern" : EMIT_RETRO_PAD_MODERN(1) ? (options.retropad_layout = "snes" : EMIT_RETRO_PAD_SNES(1) ? EMIT_RETRO_PAD(1))),EMIT_RETRO_PAD_MOUSE(1),
                               ^
Makefile:493: recipe for target 'src/mame2003/mame2003.o' failed
make: *** [src/mame2003/mame2003.o] Error 1
make: *** Waiting for unfinished jobs....
```

Looks like Tang Tang has to wait a bit more. :(

**Comment by markwkidd** [comment]
there's a little problem with dankcushion's controls PR - I didn't realize it didn't compile. fixing now, will be running again soon

**Comment by markwkidd** [comment]
@SapphireTactician should be ok now

**Comment by SapphireTactician** [comment]
@markwkidd Still no joy

```
src/vidhrdw/esd16_vidhrdw.c:44:1: error: C++ style comments are not allowed in ISO C90
 //extern data16_t *head_unknown1;
 ^
src/vidhrdw/esd16_vidhrdw.c:44:1: error: (this will be reported only once per input file)
Makefile:493: recipe for target 'src/vidhrdw/esd16_vidhrdw.o' failed
make: *** [src/vidhrdw/esd16_vidhrdw.o] Error 1
```

**Comment by markwkidd** [comment]
@SapphireTactician one more time now... I think things are straightened up now with this new driver

**Comment by SapphireTactician** [comment]
Building now, should only take about 7 minutes on this old dinosaur.

**Comment by markwkidd** [comment]
@arcadez I don't think we necessarily need this game's volume mixer in the
GUI if there is a way to hardcode a good values for BGM. Is that possible?

**Comment by SapphireTactician** [comment]
@markwkidd Can try using the volumes in the screenshot. The highlighted one is BGM, the one below is SFX. Surprised the SFX are so loud at that low volume. I don't recall the starting values but I can probably get them by deleting the NVRAM or game options.

**Comment by arcadez** [comment]
@markwkidd it would be easy to raise up the volume level for the YM3812 sound it's at 40
just now it can go as high as 100.............

static struct YM3812interface esd16_ym3812_intf =
{
	1,
	4000000,	/* ? */
	{ 100 },
	{  0 },		/* IRQ Line */
};

All the ESD16 games share it however so what might be good for Tang Tang might mess up
the sound balance in the other games that's something maybe @SapphireTactician might
like to test for..?? tell me a good balance to strike that works for all of the games then
i'll make the change in the driver.

**Comment by markwkidd** [comment]
@SapphireTactician I checked the source file `esd16` and it looks like 10 games share this driver. The question is: do they all have the ***same*** problem with background music too low? If so then the fix can be applied to everyone at once.

Would you consider expanding your process to try to investigate the games BGM volume? Here is where they are listed: https://github.com/libretro/mame2003-plus-libretro/blob/master/src/drivers/esd16.c#L938-L954

**Comment by SapphireTactician** [comment]
@markwkidd I tested out the other games, these are my findings:

The BGM volume appears to be the same or similar for all the `esd16` games, however the SFX volume is significantly louder in Tang Tang than the other `esd16` games. The other games' SFX volume varies slightly but generally sounds about equal to or slightly louder than the BGM volume.

**Comment by markwkidd** [comment]
Would you say that the volume for the games in these drivers is generally in line with games from other drivers?

In other words, do you think that the SFX in Tang Tang is the outlier, with all other sound in this driver about where it needs to be?

**Comment by SapphireTactician** [comment]
The SFX in general for the ESD games seems to be a bit higher to me on average than the rest of my collection, but Tang Tang is the worst offender in the regard by a noticeable margin imo. So in that sense, Tang Tang is an outlier among outliers.

Part of me would almost want to suggest volume sliders like the MAME 2010 core menu does, as even outside of this driver, there are other outliers. `toypop` by Namco for example has much louder BGM in comparison to the rest of my collection for example.

**Comment by arcadez** [comment]
MAME78 does support volume slider controls is there a problem with them in this core..??
i assume there must be or you would have just balanced out the bg music and sfx yourself
so that just leaves us one option then create a new sound table just for tang tang i'll do this
later on

**Comment by markwkidd** [comment]
I think it's going to be as easy as making a struct unique to tangtang in
the driver with the right volume level for the SFX or the BGM. Then when
it's tangtang loading,  have the driver pass that tangtang struct in to the
sound device rather than the generic struct.

Once we hear back on my question I think I can code up this idea in a
minute or two. I may be wrong but I think we can just fix this in the
driver transparently by setting the mixer different for just tangtang...

**Comment by SapphireTactician** [comment]
@markwkidd @arcadez commit 96fad41 broke Tang Tang

Relevant Terminal Output

```
[libretro INFO] [MAME 2003+] Total MAME drivers: 4814. Matched game driver: [tangtang].
[INFO] Environ SYSTEM_DIRECTORY: "/home/drew/.config/retroarch/system".
[INFO] Environ SET_ROTATION: 0
[libretro INFO] [MAME 2003+] Succesfully loaded ROMs.
[libretro ERROR] cpunum_get_localtime() called for invalid cpu num!
[INFO] Environ SET_PIXEL_FORMAT: RGB565.
[libretro INFO] [MAME 2003+] game bpp: [16], system bpp: [16], color format [RGB565] : SUPPORTED, enabling it.
[libretro ERROR] [MAME 2003+] Unable to start audio emulation
[ERROR] Failed to load content.
```

**Comment by markwkidd** [comment]
I do enjoying the new logging system output for mame2003-plus. Nice to see it more or less working at least.

I'll let @arcadez take a look at the error itself before I jump into this one.

**Comment by arcadez** [comment]
The game is likely getting confused around what sound table to load i'll just create a whole
new machine start for Tang Tang

**Comment by SapphireTactician** [comment]
>MAME78 does support volume slider controls is there a problem with them in this core..??
i assume there must be or you would have just balanced out the bg music and sfx yourself
so that just leaves us one option then create a new sound table just for tang tang i'll do this
later on

@arcadez Yes, volume sliders don't exist on this core for as long as I can remember. Ditto with the regular MAME 2003 core; unless they are only present on specific games and not every game? Do you want me to file that as a separate issue?

As far as Tang Tang is concerned, the PR helps the BGM volume considerably, thank you. My only remaining suggestion would be to also lower the SFX by a bit if it's not a chore to do as I feel that'd be a good balance.

Once that's taken care of, we're pretty much done with this issue.

**Comment by SapphireTactician** [comment]
Everything seems good here now. Thank you so much for all of this work!!!! This game has a lot of sentimental value to me.
